pg_analytics/src/hooks/udf.rs

```diff
@@ -0,0 +1,223 @@
+use pgrx::*;
+use std::sync::Arc;
+
+use deltalake::datafusion::arrow::datatypes::DataType;
+use deltalake::datafusion::common::DataFusionError;
+use deltalake::datafusion::common::ScalarValue;
+use deltalake::datafusion::logical_expr::{create_udf, ColumnarValue, Volatility};
+use std::cmp::max;
+use std::ffi::{CStr, CString};
+
+use crate::datafusion::session::Session;
+use crate::errors::ParadeError;
+use crate::types::array::IntoArrowArray;
+use crate::types::datatype::{ArrowDataType, PgAttribute, PgTypeMod};
+use crate::types::datum::GetDatum;
+
+// NOTE: because we don't use argtypes yet (see TODO below), using this function on overloaded functions WILL
+//       throw a postgres error
+unsafe fn func_oid_from_signature(
+    funcname: &str,
+    _argtypes: *mut pg_sys::Oid,
+) -> Result<pg_sys::Oid, ParadeError> {
+    let cstr = CString::new(funcname)?;
+
+    let funcname_list: *mut pg_sys::List;
+    #[cfg(any(feature = "pg12", feature = "pg13", feature = "pg14", feature = "pg15"))]
+    {
+        funcname_list = unsafe { pg_sys::stringToQualifiedNameList(cstr.as_ptr()) };
+    }
+    #[cfg(feature = "pg16")]
+    {
+        funcname_list =
+            unsafe { pg_sys::stringToQualifiedNameList(cstr.as_ptr(), std::ptr::null_mut()) };
+    }
+
+    // TODO: Unless we can guarantee the exact matches for arg Oids, we'll have to assume that only one function
+    // with this name exists. Non-lossy type conversion will be necessary. When non-lossy type conversion
+    // is implemented, we can turn the null_mut into argtypes.
+    Ok(pg_sys::LookupFuncName(
+        funcname_list,
+        -1,
+        std::ptr::null_mut(),
+        true,
+    ))
+}
+
+unsafe fn func_list_from_name(funcname: &str) -> Result<pg_sys::FuncCandidateList, ParadeError> {
+    let cstr = CString::new(funcname)?;
+
+    let funcname_list: *mut pg_sys::List;
+    #[cfg(any(feature = "pg12", feature = "pg13", feature = "pg14", feature = "pg15"))]
+    {
+        funcname_list = unsafe { pg_sys::stringToQualifiedNameList(cstr.as_ptr()) };
+    }
+    #[cfg(feature = "pg16")]
+    {
+        funcname_list =
+            unsafe { pg_sys::stringToQualifiedNameList(cstr.as_ptr(), std::ptr::null_mut()) };
+    }
+
+    Ok(pg_sys::FuncnameGetCandidates(
+        funcname_list,
+        -1,
+        std::ptr::null_mut(),
+        false,
+        false,
+        #[cfg(any(feature = "pg14", feature = "pg15", feature = "pg16"))]
+        false,
+        true,
+    ))
+}
+
+unsafe fn udf_datafusion(args: &[ColumnarValue]) -> Result<ColumnarValue, DataFusionError> {
```

Comment:
There's a lot of loops and conditionals in this function. Because the logic is complex, it might be a good place to do unit tests rather then trying to target all the conditions with our sqlx integration tests.

Reply:
I disagree with this. The loops are all part of reading in arguments and there really aren't a finite number of conditions to be tested with these loops. The sqlx integration test should be sufficient for testing functionality.